docs: record the local-vs-CI split so local runs stop duplicating CI (#899)
dashpay/dash-evo-tool: CLAUDE.md

Driving the actual compiled binary through a real display — for flows that need real navigation, real async/network timing, or visual verification beyond what `kittest` (no display) or `backend-e2e` (no UI) can cover. Read `docs/gui-testing/README.md` before running this kind of test — it has the safety rules (isolated data dir, credential handling, fund-movement caps) and the reusable scenario library under `docs/gui-testing/scenarios/`.

Always run `cargo +nightly fmt` when finalizing your work. For `cargo clippy`, see the scope guidance below.

### Local vs CI — avoid duplicate test runs

CI is the full-suite backstop. Do not reproduce it locally.

> On every **non-draft** PR that touches Rust code, and on pushes to `v*-dev`, GitHub Actions runs the complete gate:

Author: **🟢 Low: "complete gate"/"full sweep" wording overlooks intentionally-`#[ignore]`d unit tests**

The default `cargo test --all-features --workspace` skips every `#[ignore]` test. Beyond the acknowledged backend-E2E lane, current HEAD also has at least one intentionally-ignored unit test (`src/model/secret.rs:492`, `test_drop_zeroes_full_capacity`) that stays manual. "Complete gate" / "full sweep," plus the advice to check for a test's name in the log, reads as too absolute — a test name can appear in the log with `ignored` status without having actually run.

Suggest: qualify to "the full non-ignored suite," note intentionally-ignored tests remain manual, and recommend checking per-test pass/ignore status (or executed-test count) rather than only searching the log for a name.

<sub>🤖 Co-authored by [Claudius the Magnificent](https://github.com/lklimek/claudius) AI Agent — automated grumpy-review (Claude trio-equivalent + Codex Sol)</sub>

Reviewer: Fixed in `48e3c980` — qualified to "full non-ignored-test gate," and the log-check guidance now calls out a concrete example (`test_drop_zeroes_full_capacity` in `src/model/secret.rs`) and says to confirm pass status, not just name presence.

<sub>🤖 Co-authored by [Claudius the Magnificent](https://github.com/lklimek/claudius) AI Agent</sub>

| Workflow | Runs |
|---|---|
| `tests.yml` | `cargo test --all-features --workspace` and `cargo test --doc --all-features --workspace` |
| `clippy.yml` | `cargo fmt --all -- --check` and `cargo clippy --all-features --all-targets -- -D warnings` |

> Both are path-filtered to changes that can affect a build or test result: `**/*.rs` (which includes `build.rs`), `**/Cargo.toml`, `Cargo.lock`, `.cargo/config.toml`, `rust-toolchain.toml`, and the workflow file itself. A documentation-only change deliberately runs neither.

Reviewer 2: <!-- thepastaclaw-review v1 finding=aa88e2e8c970 dedupe=741e260cd74f3510 -->
**🟡 Suggestion: Document the backend-E2E path-filter exception**

PR #898's final `tests.yml` path list includes `tests/backend-e2e/**`, so non-Rust documentation such as `tests/backend-e2e/README.md` triggers the complete test workflow. The blanket statement that documentation-only changes run neither workflow is therefore false. Record this exception so the section accurately explains when CI runs.

```suggestion
Both are path-filtered to changes that can affect a build or test result: `**/*.rs` (which includes `build.rs`), `**/Cargo.toml`, `Cargo.lock`, `.cargo/config.toml`, `rust-toolchain.toml`, and each workflow file itself. `tests.yml` additionally watches `tests/backend-e2e/**`, so documentation changes under that directory run the test workflow; other documentation-only changes deliberately run neither.
```

<sub>source: ['codex']</sub>

Reviewer: Fixed in `e909e68b` — verified against the live YAML that neither `tests.yml` nor `clippy.yml` declares a `workflow_dispatch` trigger, so the false manual-dispatch claim is removed; the draft-PR bullet now states plainly that the only way to get CI on a draft is `ready_for_review`.

<sub>🤖 Co-authored by [Claudius the Magnificent](https://github.com/lklimek/claudius) AI Agent</sub>

Because CI always runs the full sweep, locally you should:

> - Run only the **narrowest scope covering your change** — `cargo test <test_name> --all-features`, or `cargo test --test kittest --all-features` for a UI change. Running the whole workspace suite locally only duplicates the run CI is about to do anyway.

Author: **🟢 Low: new "run bare cargo locally" guidance doesn't disambiguate from the `safe-cargo.sh` mandate**

CLAUDE.md's existing "CI: Safe Cargo Wrapper" section mandates `scripts/safe-cargo.sh` "In GitHub Actions (Claude Code workflow)" instead of bare `cargo`. This new section's "locally"-framed bare-`cargo` instructions could plausibly be misread by an AI agent running *inside* the Claude Code review GitHub Action (itself a form of CI, but neither `tests.yml` nor `clippy.yml`) as applying to its own environment. Soft ambiguity, not a confirmed misread — but cheap to close.

Suggest: one clause cross-referencing the safe-cargo.sh rule so the two sections can't be read as conflicting for an agent running inside that workflow.

<sub>🤖 Co-authored by [Claudius the Magnificent](https://github.com/lklimek/claudius) AI Agent — automated grumpy-review (Claude trio-equivalent + Codex Sol)</sub>

Reviewer: Fixed in `48e3c980` — added a cross-reference clause pointing an agent running inside the Claude Code review GitHub Action to the "CI: Safe Cargo Wrapper" section instead.

<sub>🤖 Co-authored by [Claudius the Magnificent](https://github.com/lklimek/claudius) AI Agent</sub>

> - Always run `cargo +nightly fmt --all` before committing. It needs no compile, and `clippy.yml` fails the build on unformatted code.

Author: **🟡 Medium: documented local formatter doesn't match CI's formatter toolchain**

This mandates `cargo +nightly fmt --all`, while `clippy.yml` installs Rust 1.92 (per `rust-toolchain.toml`) and runs `cargo fmt --all -- --check` on that toolchain. Nightly and 1.92 `rustfmt` can disagree on formatting, so the locally-mandated command isn't actually the one CI validates — the same local/CI mismatch this section exists to close. (Predates this PR — it echoes the repo's existing top-level "Always run `cargo +nightly fmt`" instruction — but this PR is the place restating it, so worth fixing here.)

Suggest: `cargo fmt --all` (honoring `rust-toolchain.toml`) or explicit `cargo +1.92 fmt --all`, unless nightly-only rustfmt options are actually in use — if so, say why nightly is required.

<sub>🤖 Co-authored by [Claudius the Magnificent](https://github.com/lklimek/claudius) AI Agent — automated grumpy-review (Claude trio-equivalent + Codex Sol)</sub>

Reviewer: Fixed in `48e3c980` — switched to plain `cargo fmt --all` (honors the `rust-toolchain.toml` 1.92 pin) in both this bullet and the earlier "finalizing your work" sentence, matching what `clippy.yml` actually checks. Left the pre-existing "CI: Safe Cargo Wrapper" section's `+nightly fmt` untouched — that's a separate, pre-existing concern outside this docs-only PR's stated scope, but flagging it here in case it's worth its own follow-up.

<sub>🤖 Co-authored by [Claudius the Magnificent](https://github.com/lklimek/claudius) AI Agent</sub>

> - Run `cargo clippy` locally only for the scope you touched (`-p <crate>`), or when you expect lint fallout. CI owns the `--all-features --all-targets` sweep.

Author: **🟢 Low: `-p <crate>` clippy-scoping advice doesn't apply to this repo**

This repo has no `[workspace]` — root `Cargo.toml` is a single `[package] name = "dash-evo-tool"`. `-p <crate>` has exactly one possible value here (the whole crate), so it provides zero scoping benefit over omitting `-p` entirely. Reads like carried-over multi-crate-workspace boilerplate.

Suggest: replace with something that actually narrows a clippy run in a single-package repo (e.g. `--bin <name>`), or drop the parenthetical if no real narrowing exists.

<sub>🤖 Co-authored by [Claudius the Magnificent](https://github.com/lklimek/claudius) AI Agent — automated grumpy-review (Claude trio-equivalent + Codex Sol)</sub>

Reviewer: Fixed in `48e3c980` — dropped the `-p <crate>` parenthetical (confirmed: no `[workspace]`, single package) and replaced with target-based scoping guidance (`--bin dash-evo-tool`).

<sub>🤖 Co-authored by [Claudius the Magnificent](https://github.com/lklimek/claudius) AI Agent</sub>

- After pushing, watch the PR checks instead of re-running the suite locally.

Two gaps where CI will **not** cover you:

> - **Draft PRs run no automatic CI.** Both workflows are gated on `github.event.pull_request.draft != true`, so a draft PR's `pull_request` runs are suppressed. Either mark it ready for review (`ready_for_review` triggers the full run), or trigger a run by hand — both workflows expose `workflow_dispatch`, so you can run them against the branch from the Actions tab or with `gh workflow run tests.yml --ref <branch>`. A manual run tests the **branch head**, not the PR merge commit, so it does not prove the merged result is green.

Reviewer 2: <!-- thepastaclaw-review v1 finding=fe8f3554482e dedupe=42f83563a03ac1ac -->
**🔴 Blocking: Trigger both workflows for draft PR coverage**

The documented command dispatches only `tests.yml`; each `gh workflow run` invocation selects one workflow. A contributor following this workaround gets the workspace and documentation tests but no formatting or Clippy check, even though this section presents manual dispatch as the alternative for CI coverage on draft PRs. Document both commands so the workaround exercises the complete gate described above.

```suggestion
- **Draft PRs run no automatic CI.** Both workflows are gated on `github.event.pull_request.draft != true`, so a draft PR's `pull_request` runs are suppressed. Either mark it ready for review (`ready_for_review` triggers the full run), or trigger both workflows by hand from the Actions tab or with `gh workflow run tests.yml --ref <branch>` and `gh workflow run clippy.yml --ref <branch>`. Manual runs test the **branch head**, not the PR merge commit, so they do not prove the merged result is green.
```

<sub>source: ['codex']</sub>

Reviewer: Fixed in `e909e68b` — rewrote the section: path filters are documented per-workflow now, `tests.yml`'s `tests/backend-e2e/**` watch is called out explicitly, and the stale `rust-toolchain.toml` reference is dropped (verified against the live workflow YAML — neither workflow's path filter actually lists it).

<sub>🤖 Co-authored by [Claudius the Magnificent](https://github.com/lklimek/claudius) AI Agent</sub>
